File: csharp/TheSalem/TheSalem/RoleAlignment.cs
using System;
using static TheSalem.Alignment;
using static TheSalem.Faction;

namespace TheSalem
{
    public record RoleAlignment(Faction Faction, Alignment Alignment) : IRoleSlot
    {
        #region Available role alignments
        public static readonly RoleAlignment Any = new();

        public static readonly RoleAlignment TownAny = new(Town);
        public static readonly RoleAlignment MafiaAny = new(Mafia);
        public static readonly RoleAlignment CovenAny = new(Coven);
        public static readonly RoleAlignment NeutralAny = new(Neutral);

        public static readonly RoleAlignment TownInvestigative = new(Town, Investigative);
        public static readonly RoleAlignment TownKilling = new(Town, Killing);
        public static readonly RoleAlignment TownProtective = new(Town, Protective);
        public static readonly RoleAlignment TownSupport = new(Town, Support);

        public static readonly RoleAlignment MafiaDeception = new(Mafia, Deception);
        public static readonly RoleAlignment MafiaKilling = new(Mafia, Killing);
        public static readonly RoleAlignment MafiaSupport = new(Mafia, Support);

        public static readonly RoleAlignment CovenEvil = new(Coven, Evil);

        public static readonly RoleAlignment NeutralBenign = new(Neutral, Benign);
        public static readonly RoleAlignment NeutralKilling = new(Neutral, Killing);
        public static readonly RoleAlignment NeutralEvil = new(Neutral, Evil);
        public static readonly RoleAlignment NeutralChaos = new(Neutral, Chaos);
        #endregion

        public RoleAlignment()
            : this(Faction.Any) { }
        public RoleAlignment(Alignment alignment)
            : this(Faction.Any, alignment) { }
        public RoleAlignment(Faction faction)
            : this(faction, Alignment.Any) { }

        public Role GenerateRandomRole(RoleCollection availableRoles, Random random)
        {
            var roles = availableRoles[this];
            return RoleInstancePool.Instance[roles[random.Next(roles.Length)]];
        }

        /// <summary>Determines whether this role alignment matches the given filter.</summary>
        /// <param name="alignment">The alignent filter to check whether it matches this alignment.</param>
        /// <returns><see langword="true"/> if the non-any fields of the alignment filter are equal to this alignment, otherwise <see langword="false"/>.</returns>
        public bool MatchesFilter(RoleAlignment alignment)
        {
            if (alignment.Faction != Faction.Any)
                if (alignment.Faction != Faction)
                    return false;

            if (alignment.Alignment != Alignment.Any)
                if (alignment.Alignment != Alignment)
                    return false;

            return true;
        }

        public override string ToString()
        {
            if (this == Any)
                return "Any";

            return $"{Faction} {Alignment}";
        }
        public override int GetHashCode() => ((int)Faction) | ((int)Alignment << 3);
    }
}

File: java/src/main/java/org/salem/RoleAlignment.java
package org.salem;

public class RoleAlignment implements RoleListEntry
{
    // Available role alignments
    public static final RoleAlignment Any = new RoleAlignment();

    public static final RoleAlignment TownAny = new RoleAlignment(Faction.Town);
    public static final RoleAlignment MafiaAny = new RoleAlignment(Faction.Mafia);
    public static final RoleAlignment CovenAny = new RoleAlignment(Faction.Coven);
    public static final RoleAlignment NeutralAny = new RoleAlignment(Faction.Neutral);
    
    public static final RoleAlignment TownInvestigative = new RoleAlignment(Faction.Town, Alignment.Investigative);
    public static final RoleAlignment TownKilling = new RoleAlignment(Faction.Town, Alignment.Killing);
    public static final RoleAlignment TownProtective = new RoleAlignment(Faction.Town, Alignment.Protective);
    public static final RoleAlignment TownSupport = new RoleAlignment(Faction.Town, Alignment.Support);

    public static final RoleAlignment MafiaDeception = new RoleAlignment(Faction.Mafia, Alignment.Deception);
    public static final RoleAlignment MafiaKilling = new RoleAlignment(Faction.Mafia, Alignment.Killing);
    public static final RoleAlignment MafiaSupport = new RoleAlignment(Faction.Mafia, Alignment.Support);

    public static final RoleAlignment CovenEvil = new RoleAlignment(Faction.Coven, Alignment.Evil);

    public static final RoleAlignment NeutralBenign = new RoleAlignment(Faction.Neutral, Alignment.Benign);
    public static final RoleAlignment NeutralKilling = new RoleAlignment(Faction.Neutral, Alignment.Killing);
    public static final RoleAlignment NeutralEvil = new RoleAlignment(Faction.Neutral, Alignment.Evil);
    public static final RoleAlignment NeutralChaos = new RoleAlignment(Faction.Neutral, Alignment.Chaos);
    
    public final Faction RoleFaction;
    public final Alignment RoleAlignment;

    public RoleAlignment()
    {
        this(Faction.Any, Alignment.Any);
    }
    public RoleAlignment(Faction faction)
    {
        this(faction, Alignment.Any);
    }
    public RoleAlignment(Faction faction, Alignment alignment)
    {
        RoleFaction = faction;
        RoleAlignment = alignment;
    }

    @Override
    public String toString()
    {
        if (RoleFaction == Faction.Any)
            return "Any";

        return RoleFaction + " " + RoleAlignment;
    }
    @Override
    public int hashCode()
    {
        return (RoleFaction.ordinal()) | (RoleAlignment.ordinal() << 3);
    }
}
